File: androidsupport/src/main/java/com/tq/libs/service/MessageHandler.java
/*
 * ******************************************************************************
 *  Copyright Ⓒ 2016. TrinhQuan. All right reserved
 *  Author: TrinhQuan. Created on 2016/12/20
 *  Contact: trinhquan.171093@gmail.com
 * ******************************************************************************
 */

package com.tq.libs.service;

import android.os.Handler;
import android.os.Message;

import com.tq.libs.common.Handleable;

import java.lang.ref.WeakReference;

public final class MessageHandler extends Handler {

    private final WeakReference<Handleable> handleableWeakReference;

    public MessageHandler(Handleable handleable) {
        handleableWeakReference = new WeakReference<>(handleable);
    }

    @Override
    public void handleMessage(Message msg) {
        super.handleMessage(msg);
        Handleable handleable = handleableWeakReference.get();
        if (handleable != null) {
            handleable.handleMessage(msg);
        }
    }
}

File: kotlinandroidsupport/src/main/java/com/tq/klibs/service/MessageHandler.kt
/*
 * ******************************************************************************
 *  Copyright Ⓒ 2016. TrinhQuan. All right reserved
 *  Author: TrinhQuan. Created on 2016/12/20
 *  Contact: trinhquan.171093@gmail.com
 * ******************************************************************************
 */

package com.tq.klibs.service

import android.os.Handler
import android.os.Message

import com.tq.klibs.common.Handleable

import java.lang.ref.WeakReference

class MessageHandler(handleable: Handleable) : Handler() {

    private val handleableWeakReference: WeakReference<Handleable>

    init {
        handleableWeakReference = WeakReference(handleable)
    }

    override fun handleMessage(msg: Message) {
        super.handleMessage(msg)
        val handleable = handleableWeakReference.get()
        handleable?.handleMessage(msg)
    }
}
